File: backend/testing_tooling/uploadtests.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"math/rand"
	"mime/multipart"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// ── ANSI color codes ──────────────────────────────────────────────────────────

const (
	reset   = "\033[0m"
	bold    = "\033[1m"
	dim     = "\033[2m"
	red     = "\033[31m"
	green   = "\033[32m"
	yellow  = "\033[33m"
	blue    = "\033[34m"
	magenta = "\033[35m"
	cyan    = "\033[36m"
	white   = "\033[37m"
	bgBlue  = "\033[44m"
	bgCyan  = "\033[46m"
)

// ── Random data pools ─────────────────────────────────────────────────────────

var (
	firstNames  = []string{"Oliver", "Phoebe", "Marcus", "Ingrid", "Tariq", "Yuki", "Soren", "Amara", "Declan", "Priya"}
	lastNames   = []string{"Nakamura", "Osei", "Lindqvist", "Ferrara", "Patel", "Kowalski", "Okafor", "Reyes", "Svensson", "Mbeki"}
	employers   = []string{"Horizon Robotics", "Starfall Media", "Ironclad Materials", "Vivant Health", "Obsidian Logistics", "Luminary Tech", "Verdant Farms", "Nexus Analytics", "Solaris Energy", "Phalanx Security"}
	adjusters   = []string{"T. Hargrove", "M. Delacroix", "A. Fujimoto", "R. Oduya", "S. Bergmann", "C. Abramowitz", "D. Kazakov", "F. Osei-Mensah", "L. Cartwright", "P. Iyer"}
	claimTypes  = []string{"Workers Comp", "Liability", "Property", "Medical", "Disability", "Auto", "Product Liability", "Environmental"}
	supportLvls = []string{"Full Support", "Partial", "Minimal", "Psychiatric", "Physical Therapy", "None", "Pending Review"}
	jurisdicts  = []string{"California", "New York", "Texas", "Florida", "Illinois", "Washington", "Colorado", "Georgia", "Ohio", "Michigan"}
	filePairs   = []struct{ path, label string }{
		{"two.png", "two.png"},
		{"lenna.jpg", "lenna.jpg"},
		{"test.pdf", "test.pdf"},
		{"CrossDocTesting", "CrossDocTesting"},
	}
	statusMessages = []string{
		"Dispatching request…",
		"Knocking on the server door…",
		"Firing off multipart form…",
		"Launching claim into the void…",
		"Sending bytes across the wire…",
	}
)

func rnd(pool []string) string { return pool[rand.Intn(len(pool))] }

func randDate() string {
	start := time.Date(2024, 1, 1, 0, 0, 0, 0, time.UTC)
	days := rand.Intn(730)
	return start.AddDate(0, 0, days).Format("2006-01-02")
}

func randClaimNum() string { return fmt.Sprintf("CLM-%05d", rand.Intn(90000)+10000) }
func randPolicyNum() string {
	return fmt.Sprintf("POL-%06d", rand.Intn(900000)+100000)
}
func randActsID(i int) string { return fmt.Sprintf("ACTS_%03d", i) }

// ── Pretty printing helpers ───────────────────────────────────────────────────

func header() {
	w := 60
	line := strings.Repeat("─", w)
	fmt.Printf("\n%s%s╔%s╗%s\n", bold, cyan, strings.Repeat("═", w), reset)
	fmt.Printf("%s%s║%s  %-*s%s%s║%s\n", bold, cyan, white, w-2, "  🧪  CLAIM UPLOAD TEST SUITE", cyan, bold, reset)
	fmt.Printf("%s%s╚%s╝%s\n\n", bold, cyan, strings.Repeat("═", w), reset)
	_ = line
}

func sectionBanner(n int, total int, label string) {
	bar := fmt.Sprintf("[%d/%d]", n, total)
	fmt.Printf("%s%s  %s %-45s%s\n", bold, yellow, bar, label, reset)
	fmt.Printf("%s%s%s\n", dim, strings.Repeat("·", 60), reset)
}

func fieldLine(key, value string) {
	fmt.Printf("  %s%-20s%s %s%s%s\n", dim, key, reset, cyan, value, reset)
}

func statusLine(msg string) {
	fmt.Printf("\n  %s⟳  %s%s\n", magenta, msg, reset)
}

func successLine(code int, took time.Duration) {
	color := green
	icon := "✓"
	if code >= 400 {
		color = red
		icon = "✗"
	} else if code >= 300 {
		color = yellow
		icon = "↪"
	}
	fmt.Printf("  %s%s%s  HTTP %s%d%s  %s(%s)%s\n\n",
		bold, color, icon, bold, code, reset, dim, took.Round(time.Millisecond), reset)
}

func errorLine(err error) {
	fmt.Printf("  %s✗  ERROR: %v%s\n\n", red, err, reset)
}

func separator() {
	fmt.Printf("%s%s%s\n", dim, strings.Repeat("─", 60), reset)
}

func summary(passed, failed int, total time.Duration) {
	fmt.Printf("\n%s%s╔%s╗%s\n", bold, cyan, strings.Repeat("═", 60), reset)
	fmt.Printf("%s%s║%s  %-58s%s%s║%s\n", bold, cyan, white, "  RESULTS", cyan, bold, reset)
	fmt.Printf("%s%s╠%s╣%s\n", bold, cyan, strings.Repeat("═", 60), reset)
	fmt.Printf("%s%s║%s  %-58s%s%s║%s\n", bold, cyan, green,
		fmt.Sprintf("  ✓  Passed : %d", passed), cyan, bold, reset)
	fmt.Printf("%s%s║%s  %-58s%s%s║%s\n", bold, cyan, red,
		fmt.Sprintf("  ✗  Failed : %d", failed), cyan, bold, reset)
	fmt.Printf("%s%s║%s  %-58s%s%s║%s\n", bold, cyan, dim,
		fmt.Sprintf("  ⏱  Total  : %s", total.Round(time.Millisecond)), cyan, bold, reset)
	fmt.Printf("%s%s╚%s╝%s\n\n", bold, cyan, strings.Repeat("═", 60), reset)
}

// ── HTTP helpers ──────────────────────────────────────────────────────────────

func doGET(url string) (int, error) {
	resp, err := http.Get(url)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	return resp.StatusCode, nil
}

func doUpload(url string, fields map[string]string, filePath string) (int, error) {
	var buf bytes.Buffer
	w := multipart.NewWriter(&buf)

	for k, v := range fields {
		if err := w.WriteField(k, v); err != nil {
			return 0, err
		}
	}

	// file field (skip if path empty or file missing)
	if filePath != "" {
		f, err := os.Open(filePath)
		if err == nil {
			defer f.Close()
			fw, err := w.CreateFormFile("file", filepath.Base(filePath))
			if err != nil {
				return 0, err
			}
			io.Copy(fw, f)
		}
		// if file doesn't exist we just skip it — testing infra may not have them
	}

	w.Close()

	req, err := http.NewRequest("POST", url, &buf)
	if err != nil {
		return 0, err
	}
	req.Header.Set("Content-Type", w.FormDataContentType())

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	return resp.StatusCode, nil
}

// ── Test cases ────────────────────────────────────────────────────────────────

type TestCase struct {
	label  string
	run    func() (int, error)
	fields map[string]string
}

func buildCases(base string) []TestCase {
	cases := make([]TestCase, 0, len(filePairs)+1)

	// GET health check
	cases = append(cases, TestCase{
		label:  "GET /  (health check)",
		fields: map[string]string{"url": base + "/"},
		run:    func() (int, error) { return doGET(base + "/") },
	})

	// One upload per file
	for i, fp := range filePairs {
		fp := fp // capture
		idx := i + 1
		fname := fmt.Sprintf("%s %s", rnd(firstNames), rnd(lastNames))
		fields := map[string]string{
			"claim_number":    randClaimNum(),
			"claimant_name":   fname,
			"date_of_injury":  randDate(),
			"employer":        rnd(employers),
			"adjuster":        rnd(adjusters),
			"support":         rnd(supportLvls),
			"claim_type":      rnd(claimTypes),
			"jurisdiction":    rnd(jurisdicts),
			"policy_number":   randPolicyNum(),
			"acts_id":         randActsID(idx),
			"data":            fmt.Sprintf("run-%d-data-%04x", idx, rand.Intn(0xffff)),
		}
		fCopy := make(map[string]string, len(fields))
		for k, v := range fields {
			fCopy[k] = v
		}
		cases = append(cases, TestCase{
			label:  fmt.Sprintf("POST /upload  [%s]", fp.label),
			fields: fCopy,
			run: func() (int, error) {
				return doUpload(base+"/upload", fCopy, fp.path)
			},
		})
	}
	return cases
}

// ── main ──────────────────────────────────────────────────────────────────────

func main() {
	rand.Seed(time.Now().UnixNano())

	base := "http://localhost:8080"
	if len(os.Args) > 1 {
		base = strings.TrimRight(os.Args[1], "/")
	}

	header()
	fmt.Printf("  %sTarget:%s %s%s%s\n\n", dim, reset, bold, base, reset)

	cases := buildCases(base)
	total := len(cases)
	passed, failed := 0, 0
	start := time.Now()

	for i, tc := range cases {
		sectionBanner(i+1, total, tc.label)

		// print fields (skip internal ones)
		skip := map[string]bool{"url": true}
		for _, k := range []string{
			"claim_number", "claimant_name", "date_of_injury", "employer",
			"adjuster", "support", "claim_type", "jurisdiction",
			"policy_number", "acts_id", "data",
		} {
			if v, ok := tc.fields[k]; ok && !skip[k] {
				fieldLine(k+":", v)
			}
		}

		statusLine(rnd(statusMessages))

		t0 := time.Now()
		code, err := tc.run()
		elapsed := time.Since(t0)

		if err != nil {
			errorLine(err)
			failed++
		} else {
			successLine(code, elapsed)
			if code < 400 {
				passed++
			} else {
				failed++
			}
		}

		separator()
		time.Sleep(400 * time.Millisecond)
	}

	summary(passed, failed, time.Since(start))
}

File: backend/testing_tooling/uploadTests.lua
#!/usr/bin/env lua

-- Seed the random number generator
math.randomseed(os.time())

-- ── ANSI color codes ──────────────────────────────────────────────────────────
local reset          = "\27[0m"
local bold           = "\27[1m"
local dim            = "\27[2m"
local red            = "\27[31m"
local green          = "\27[32m"
local yellow         = "\27[33m"
local magenta        = "\27[35m"
local cyan           = "\27[36m"
local white          = "\27[37m"

-- ── Random data pools ─────────────────────────────────────────────────────────
local firstNames     = { "Oliver", "Phoebe", "Marcus", "Ingrid", "Tariq", "Yuki", "Soren", "Amara", "Declan", "Priya" }
local lastNames      = { "Nakamura", "Osei", "Lindqvist", "Ferrara", "Patel", "Kowalski", "Okafor", "Reyes", "Svensson",
    "Mbeki" }
local employers      = { "Horizon Robotics", "Starfall Media", "Ironclad Materials", "Vivant Health",
    "Obsidian Logistics", "Luminary Tech", "Verdant Farms", "Nexus Analytics", "Solaris Energy", "Phalanx Security" }
local adjusters      = { "T. Hargrove", "M. Delacroix", "A. Fujimoto", "R. Oduya", "S. Bergmann", "C. Abramowitz",
    "D. Kazakov", "F. Osei-Mensah", "L. Cartwright", "P. Iyer" }
local claimTypes     = { "Workers Comp", "Liability", "Property", "Medical", "Disability", "Auto", "Product Liability",
    "Environmental" }
local supportLvls    = { "Full Support", "Partial", "Minimal", "Psychiatric", "Physical Therapy", "None",
    "Pending Review" }
local jurisdicts     = { "California", "New York", "Texas", "Florida", "Illinois", "Washington", "Colorado", "Georgia",
    "Ohio", "Michigan" }

local filePairs      = {
    { path = "two.png",         label = "two.png" },
    { path = "lenna.jpg",       label = "lenna.jpg" },
    { path = "test.pdf",        label = "test.pdf" },
    { path = "CrossDocTesting", label = "CrossDocTesting" },
    { path = "renamedELF.txt",  label = "renamedELF.txt" },
}

local statusMessages = {
    "Dispatching request…",
    "Knocking on the server door…",
    "Firing off multipart form…",
    "Launching claim into the void…",
    "Sending bytes across the wire…",
}

-- ── Random Utility Functions ──────────────────────────────────────────────────
local function rnd(pool)
    return pool[math.random(#pool)]
end

local function randDate()
    -- 2024-01-01 basic epoch calculation
    local start_ts = os.time({ year = 2024, month = 1, day = 1, hour = 0, min = 0, sec = 0 })
    local days_seconds = math.random(0, 729) * 24 * 60 * 60
    return os.date("%Y-%m-%d", start_ts + days_seconds)
end

local function randClaimNum() return string.format("CLM-%05d", math.random(10000, 99999)) end
local function randPolicyNum() return string.format("POL-%06d", math.random(100000, 999999)) end
local function randActsID(i) return string.format("ACTS_%03d", i) end

local function getBasename(path)
    return path:match("^.+/(.+)$") or path
end

-- ── Pretty printing helpers ───────────────────────────────────────────────────
local function header()
    local w = 60
    print(string.format("\n%s%s╔%s╗%s", bold, cyan, string.rep("═", w), reset))
    print(string.format("%s%s║%s  🧪  CLAIM & FILE VERIFICATION TEST SUITE%s%s%s║%s", bold, cyan, white,
        string.rep(" ", w - 43), cyan, bold, reset))
    print(string.format("%s%s╚%s╝%s\n", bold, cyan, string.rep("═", w), reset))
end

local function sectionBanner(n, total, label)
    local bar = string.format("[%d/%d]", n, total)
    print(string.format("%s%s  %s %-45s%s", bold, yellow, bar, label, reset))
    print(string.format("%s%s%s", dim, string.rep("·", 60), reset))
end

local function fieldLine(key, value)
    print(string.format("  %s%-20s%s %s%s%s", dim, key, reset, cyan, value, reset))
end

local function statusLine(msg)
    print(string.format("\n  %s⟳  %s%s", magenta, msg, reset))
end

local function successLine(code, took_ms)
    local color = green
    local icon = "✓"
    if code >= 400 then
        color = red
        icon = "✗"
    elseif code >= 300 then
        color = yellow
        icon = "↪"
    end
    print(string.format("  %s%s%s  HTTP %s%d%s  %s(%dms)%s\n",
        bold, color, icon, bold, code, reset, dim, math.floor(took_ms + 0.5), reset))
end

local function errorLine(err)
    print(string.format("  %s✗  ERROR: %s%s\n", red, tostring(err), reset))
end

local function separator()
    print(string.format("%s%s%s", dim, string.rep("─", 60), reset))
end

local function summary(passed, failed, total_ms)
    local w = 60
    print(string.format("\n%s%s╔%s╗%s", bold, cyan, string.rep("═", w), reset))
    print(string.format("%s%s║%s  RESULTS%s%s%s║%s", bold, cyan, white, string.rep(" ", w - 9), cyan, bold, reset))
    print(string.format("%s%s╠%s╣%s", bold, cyan, string.rep("═", w), reset))

    local p_str = string.format("  ✓  Passed : %d", passed)
    print(string.format("%s%s║%s%s%s%s%s║%s", bold, cyan, green, p_str, string.rep(" ", w - #p_str), cyan, bold, reset))

    local f_str = string.format("  ✗  Failed : %d", failed)
    print(string.format("%s%s║%s%s%s%s%s║%s", bold, cyan, red, f_str, string.rep(" ", w - #f_str), cyan, bold, reset))

    local t_str = string.format("  ⏱  Total  : %dms", math.floor(total_ms + 0.5))
    print(string.format("%s%s║%s%s%s%s%s║%s", bold, cyan, dim, t_str, string.rep(" ", w - #t_str), cyan, bold, reset))

    print(string.format("%s%s╚%s╝%s\n", bold, cyan, string.rep("═", w), reset))
end

-- ── HTTP Helper Functions ─────────────────────────────────────────────────────
-- Attempt to load luasocket/luasec safely
local http = require("socket.http")
local ltn12 = require("ltn12")

local function doGET(url)
    local response_body = {}
    local res, code, _ = http.request {
        url = url,
        method = "GET",
        sink = ltn12.sink.table(response_body),
        timeout = 10
    }
    if not res then return 0, code end
    return code, nil
end

local function doUpload(url, fields, filePath)
    local boundary = "----LuaMultipartBoundary" .. string.format("%04x", math.random(0, 0xffff))
    local body_chunks = {}

    -- Add regular form fields
    for k, v in pairs(fields) do
        table.insert(body_chunks,
            string.format("--%s\r\nContent-Disposition: form-data; name=\"%s\"\r\n\r\n%s\r\n", boundary, k, v))
    end

    -- Add file payload if available
    if filePath then
        local f = io.open(filePath, "rb")
        if f then
            local content = f:read("*all")
            f:close()
            local filename = getBasename(filePath)
            table.insert(body_chunks,
                string.format(
                "--%s\r\nContent-Disposition: form-data; name=\"file\"; filename=\"%s\"\r\nContent-Type: application/octet-stream\r\n\r\n%s\r\n",
                    boundary, filename, content))
        end
    end

    table.insert(body_chunks, string.format("--%s--\r\n", boundary))
    local body = table.concat(body_chunks)

    local response_body = {}
    local res, code, _ = http.request {
        url = url,
        method = "POST",
        headers = {
            ["Content-Type"] = "multipart/form-data; boundary=" .. boundary,
            ["Content-Length"] = tostring(#body)
        },
        source = ltn12.source.string(body),
        sink = ltn12.sink.table(response_body),
        timeout = 10
    }

    if not res then return 0, code end
    return code, nil
end

-- ── Main Execution ────────────────────────────────────────────────────────────
local function main()
    local base = "http://localhost:8080"
    if arg[1] then
        base = arg[1]:gsub("/+$", "") -- Trim trailing slashes
    end

    header()
    print(string.format("  %sTarget:%s %s%s%s\n", dim, reset, bold, base, reset))

    local cases = {}

    -- Step 1: Baseline Health Check
    table.insert(cases, {
        label = "GET / (health check)",
        fields = { url = base .. "/" },
        run = function() return doGET(base .. "/") end
    })

    -- Step 2: Build validation chains for files dynamically
    for i, fp in ipairs(filePairs) do
        local fname = rnd(firstNames) .. " " .. rnd(lastNames)
        local filenameOnly = getBasename(fp.path)

        local fields = {
            claim_number   = randClaimNum(),
            claimant_name  = fname,
            date_of_injury = randDate(),
            employer       = rnd(employers),
            adjuster       = rnd(adjusters),
            support        = rnd(supportLvls),
            claim_type     = rnd(claimTypes),
            jurisdiction   = rnd(jurisdicts),
            policy_number  = randPolicyNum(),
            acts_id        = randActsID(i),
            data           = string.format("run-%d-data-%04x", i, math.random(0, 0xffff)),
        }

        -- A: Document Submission Test
        table.insert(cases, {
            label = string.format("POST /upload [%s]", fp.label),
            fields = fields,
            run = function() return doUpload(base .. "/upload", fields, fp.path) end
        })

        -- B: Metadata Verification
        table.insert(cases, {
            label = string.format("GET /files/%s/true (Verify Meta)", filenameOnly),
            fields = { file = filenameOnly, meta_flag = "true" },
            run = function() return doGET(string.format("%s/files/%s/true", base, filenameOnly)) end
        })

        -- C: Payload Download Verification
        table.insert(cases, {
            label = string.format("GET /files/%s/false (Verify File)", filenameOnly),
            fields = { file = filenameOnly, meta_flag = "false" },
            run = function() return doGET(string.format("%s/files/%s/false", base, filenameOnly)) end
        })
    end

    -- Step 3: Global Catalog Verification
    table.insert(cases, {
        label = "GET /files (Verify Index Inventory)",
        fields = { endpoint = "/files" },
        run = function() return doGET(base .. "/files") end
    })

    -- Execution Loop
    local total = #cases
    local passed, failed = 0, 0
    -- Simple clock metric approximation in seconds
    local startTime = os.clock()

    for i, tc in ipairs(cases) do
        sectionBanner(i, total, tc.label)

        for k, v in pairs(tc.fields) do
            fieldLine(k .. ":", v)
        end

        statusLine(rnd(statusMessages))

        local t0 = os.clock()
        local code, err = tc.run()
        local elapsed_ms = (os.clock() - t0) * 1000

        if code == 0 or err then
            errorLine(err or "Connection Failed")
            failed = failed + 1
        else
            successLine(code, elapsed_ms)
            if code < 400 then
                passed = passed + 1
            else
                failed = failed + 1
            end
        end

        separator()

        -- Sleep replacement using a busy wait or socket select
        local socket = require("socket")
        socket.select(nil, nil, 0.250)
    end

    local total_duration_ms = (os.clock() - startTime) * 1000
    summary(passed, failed, total_duration_ms)
end

main()
